### PR TITLE
Add ability to generate non nullable messages, convert Resource to test the change

### PR DESCRIPTION
**Pull request by bogdandrutu**
Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

### CODE REVIEW

File: cmd/pdatagen/internal/base_fields.go [hunk 1]
```diff
@@ -55,6 +55,25 @@ const accessorsMessageTestTemplate = `func Test${structName}_${fieldName}(t *tes
 	assert.EqualValues(t, generateTest${returnType}(), ms.${fieldName}())
 }`
 
+const accessorsMessageValueTemplate = `// ${fieldName} returns the ${lowerFieldName} associated with this ${structName}.
+// If no ${lowerFieldName} available, it creates an empty message and associates it with this ${structName}.
+//
+//  Empty initialized ${structName} will return "nil" ${returnType}.
+//
+// Important: This causes a runtime error if IsNil() returns "true".
```
Comment (tigrannajaryan) [inline cmd/pdatagen/internal/base_fields.go:63]: This comments no longer apply, since it is not a pointer anymore.

File: consumer/pdata/generated_resource.go [hunk 2]
```diff
@@ -66,14 +63,16 @@ func (ms Resource) Attributes() AttributeMap {
...
+}
+
 // CopyTo copies all properties from the current struct to the dest.
 func (ms Resource) CopyTo(dest Resource) {
-	if ms.IsNil() {
-		*dest.orig = nil
+	if ms.IsEmpty() {
+		*dest.orig = emptyValueResource
 		return
 	}
```
Comment (tigrannajaryan) [inline consumer/pdata/generated_resource.go:76]: Why is this needed? Isn't it sufficient to do `CopyTo` for all fields?

File: consumer/pdata/generated_resource.go [hunk 3]
```diff
@@ -66,14 +63,16 @@ func (ms Resource) Attributes() AttributeMap {
 	return newAttributeMap(&(*ms.orig).Attributes)
 }
 
+// IsEmpty returns true if the underlying data are equivalent with an empty message.
+func (ms Resource) IsEmpty() bool {
+	return ms.Attributes().IsEmpty()
+}
```
Comment (tigrannajaryan) [inline consumer/pdata/generated_resource.go:69]: Why do we need this?

File: exporter/loggingexporter/logging_exporter.go [hunk 4]
```diff
@@ -325,7 +325,7 @@ func (s *loggingExporter) pushTraceData(
 			buf.logEntry("* Nil ResourceSpans")
 			continue
 		}
-		if !rs.Resource().IsNil() {
+		if !rs.Resource().IsEmpty() {
```
Comment (tigrannajaryan) [inline exporter/loggingexporter/logging_exporter.go:328]: This check is not necessary anymore.

File: internal/processor/filterspan/filterspan.go [hunk 5]
```diff
@@ -125,7 +125,7 @@ func (mp *propertiesMatcher) MatchSpan(span pdata.Span, resource pdata.Resource,
 
 // serviceNameForResource gets the service name for a specified Resource.
 func serviceNameForResource(resource pdata.Resource) string {
-	if resource.IsNil() {
+	if resource.IsEmpty() {
 		return "<nil-resource>"
 	}
```
Comment (tigrannajaryan) [inline internal/processor/filterspan/filterspan.go:130]: I think this can also be deleted. The distinction between "<nil-resource>" and "<nil-service-name>" is artificial and unnecessary.

File: processor/attributesprocessor/attributes_trace_test.go [hunk 6]
```diff
@@ -75,7 +74,7 @@ func sortAttributes(td pdata.Traces) {
 		if rs.IsNil() {
 			continue
 		}
-		if !rs.Resource().IsNil() {
+		if !rs.Resource().IsEmpty() {
```
Comment (tigrannajaryan) [inline processor/attributesprocessor/attributes_trace_test.go:77]: Not needed, can Sort empty attributes.

File: processor/resourceprocessor/resource_processor_test.go [hunk 7]
```diff
@@ -250,7 +247,7 @@ func (tln *testLogsConsumer) ConsumeLogs(_ context.Context, ld pdata.Logs) error
 }
 
 func sortResourceAttributes(resource pdata.Resource) {
-	if resource.IsNil() {
+	if resource.IsEmpty() {
 		return
 	}
```
Comment (tigrannajaryan) [inline processor/resourceprocessor/resource_processor_test.go:252]: Not needed.

File: translator/internaldata/resource_to_oc.go [hunk 8]
```diff
@@ -79,7 +79,7 @@ func getSDKLangToOCLangCodeMap() map[string]int32 {
 }
 
 func internalResourceToOC(resource pdata.Resource) (*occommon.Node, *ocresource.Resource) {
-	if resource.IsNil() {
+	if resource.IsEmpty() {
 		return nil, nil
 	}
```
Comment (tigrannajaryan) [inline translator/internaldata/resource_to_oc.go:84]: Can be deleted or substituted by `if attrs.Len() == 0` check that we have below if want to return nil in that case.

File: translator/trace/jaeger/traces_to_jaegerproto.go [hunk 9]
```diff
@@ -104,7 +104,7 @@ func resourceSpansToJaegerProto(rs pdata.ResourceSpans) (*model.Batch, error) {
 func resourceToJaegerProtoProcess(resource pdata.Resource) *model.Process {
 
 	process := model.Process{}
-	if resource.IsNil() {
+	if resource.IsEmpty() {
 		process.ServiceName = tracetranslator.ResourceNotSet
 		return &process
 	}
```
Comment (tigrannajaryan) [inline translator/trace/jaeger/traces_to_jaegerproto.go:110]: I think this is also not necessary, it is not different from `if attrs.Len() == 0 {` case below.

File: cmd/pdatagen/internal/base_structs.go [hunk 10]
```diff
@@ -338,19 +338,91 @@ func Test${structName}_CopyTo(t *testing.T) {
...
+	*ms.orig = ${originName}{}
+}
+
+// Deprecated: Use IsEmpty instead
+func (ms ${structName}) IsNil() bool {
+	return ms.IsEmpty()
+}`
+
+const messageValueIsEmptyHeaderTemplate = `// IsEmpty returns true if the underlying data are equivalent with an empty message.
+func (ms ${structName}) IsEmpty() bool {
```
Comment (tigrannajaryan) [inline cmd/pdatagen/internal/base_structs.go:387]: I wonder if we can completely eliminate IsEmpty for value types. All the usages in the codebase that I can see I think can are very likely possible to remove safely.

File: cmd/pdatagen/internal/base_fields.go (hunk 1 above)
Comment (bogdandrutu) [reply cmd/pdatagen/internal/base_fields.go:63]: It still applies because this is the getter for a non double pointer message, but the parent struct that embeds this message can still be double pointer.

File: exporter/loggingexporter/logging_exporter.go (hunk 4 above)
Comment (bogdandrutu) [reply exporter/loggingexporter/logging_exporter.go:328]: Good call, removed

File: internal/processor/filterspan/filterspan.go (hunk 5 above)
Comment (bogdandrutu) [reply internal/processor/filterspan/filterspan.go:130]: Yes agreed. removed even the distinction.

File: processor/attributesprocessor/attributes_trace_test.go (hunk 6 above)
Comment (bogdandrutu) [reply processor/attributesprocessor/attributes_trace_test.go:77]: Done.

File: processor/resourceprocessor/resource_processor_test.go (hunk 7 above)
Comment (bogdandrutu) [reply processor/resourceprocessor/resource_processor_test.go:252]: Done.

File: translator/internaldata/resource_to_oc.go (hunk 8 above)
Comment (bogdandrutu) [reply translator/internaldata/resource_to_oc.go:84]: Done.

File: translator/trace/jaeger/traces_to_jaegerproto.go (hunk 9 above)
Comment (bogdandrutu) [reply translator/trace/jaeger/traces_to_jaegerproto.go:110]: Done.

File: cmd/pdatagen/internal/base_structs.go (hunk 10 above)
Comment (bogdandrutu) [reply cmd/pdatagen/internal/base_structs.go:387]: Done.

File: consumer/pdata/generated_resource.go (hunk 2 above)
Comment (bogdandrutu) [reply consumer/pdata/generated_resource.go:76]: Removed.

File: consumer/pdata/generated_resource.go (hunk 3 above)
Comment (bogdandrutu) [reply consumer/pdata/generated_resource.go:69]: Done.

**Comment by bogdandrutu** [comment]
@tigrannajaryan PTAL

**Comment by tigrannajaryan** [comment]
Conceptually LGTM. Should we start with slices first though since that will give us the most gains?

**Comment by bogdandrutu** [comment]
@tigrannajaryan can you clarify "Conceptually LGTM. Should we start with slices first though since that will give us the most gains?". Instead of changing Resource to also add support to have slices of "messageValue" and start changing some of the slices?

I think it is fine, but "Resource" also gives us something and it is still important to have one object that is not part of the slice I think.

**Comment by tigrannajaryan** [comment]
> @tigrannajaryan can you clarify "Conceptually LGTM. Should we start with slices first though since that will give us the most gains?". Instead of changing Resource to also add support to have slices of "messageValue" and start changing some of the slices?
> 
> I think it is fine, but "Resource" also gives us something and it is still important to have one object that is not part of the slice I think.

If I understand correctly one of the goals of this change is improving performance. We will not see much gains from changing the Resource from pointer to value type. The most gains can be seen from the most numerous structs, e.g. Span, which are part of InstrumentationSpans slice.

I am fine if you want to still go ahead with the Resource and use it as the playground for the upcoming changes for other structs.

**Comment by bogdandrutu** [comment]
That was my intention, to use this as a playground. Changing a slice as well would make the change to big.

Also in this PR we improved couple of other things like the service name.

**Review by tigrannajaryan — APPROVED** [review]

**Comment by tigrannajaryan** [comment]
@bogdandrutu please resolve the conflicts.
